atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/TransactionPreCommitActions.java

```diff
@@ -18,81 +18,96 @@
 
 import com.google.common.collect.Maps;
 import com.palantir.atlasdb.common.api.timelock.TimestampLeaseName;
+import com.palantir.atlasdb.transaction.api.TimestampLeaseAwareTransaction;
+import com.palantir.atlasdb.transaction.api.TimestampLeaseAwareTransaction.PreCommitAction;
+import com.palantir.atlasdb.transaction.api.Transaction;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.function.Consumer;
-import java.util.function.LongSupplier;
 import javax.annotation.concurrent.GuardedBy;
 import javax.annotation.concurrent.ThreadSafe;
 
+/**
+ * Keeps track of preCommit actions added through the course of a {@link Transaction}.
+ * Such actions are fetched via {@link #getActions()} and executed on {@link Transaction#commit()}.
+ * See {@link TimestampLeaseAwareTransaction#preCommit(TimestampLeaseName, int, PreCommitAction)} for more information.
+ */
 @ThreadSafe
 final class TransactionPreCommitActions {
 
-    static class PreCommitAction {
-        final Consumer<LongSupplier> action;
-        final int timestampCount;
+    static class PreCommitActionWrapper {
+        final PreCommitAction action;
+        final int numLeasedTimestamps;
 
-        PreCommitAction(Consumer<LongSupplier> action, int timestampCount) {
+        PreCommitActionWrapper(PreCommitAction action, int numLeasedTimestamps) {
             this.action = action;
-            this.timestampCount = timestampCount;
+            this.numLeasedTimestamps = numLeasedTimestamps;
         }
 
         @Override
         public boolean equals(Object obj) {
-            return obj instanceof PreCommitAction
-                    && timestampCount == ((PreCommitAction) obj).timestampCount
-                    && action.equals(((PreCommitAction) obj).action);
+            return obj instanceof PreCommitActionWrapper
+                    && numLeasedTimestamps == ((PreCommitActionWrapper) obj).numLeasedTimestamps
+                    && action.equals(((PreCommitActionWrapper) obj).action);
         }
 
         @Override
         public int hashCode() {
-            return action.hashCode();
+            int hash = 5381;
+            hash += (hash << 5) + action.hashCode();
+            hash += (hash << 5) + Integer.hashCode(numLeasedTimestamps);
+            return hash;
         }
```

Comment:
sorry, not sure I followed https://github.com/palantir/atlasdb/pull/7305#discussion_r1822630034 - I get that PerLeaseActions isn't immutable, but PreCommitActionWrapper is right? The fields are final (unless you plan to extend this in the future in a way that is not going to be immutable?)